### PR TITLE
First version newtonDiagnostics tool

### PR DESCRIPTION
**Pull request by TeusvanderStelt**

### CODE REVIEW

**Review by mahge — CHANGES_REQUESTED** [review]
The compilation is failing because both `newton_diagnostics.c` and `newton_diagnostics_with_debug.c` are being compiled by the CMake build system. 

I see that you have updated one of the CMakeLists files. However, we actually have two CMake compilation paths. I can understand it is a bit confusing. One of them is old and used just for MSVC compilation the other one is new/experimental and intended to be used in general.

Anyway without throwing too much information at you, the new CMake build collects source files by using a glob expression. So any source file in the folder simulation/solver will be compiled because of this
https://github.com/OpenModelica/OpenModelica/blob/1772c4b421f15d729b6d1b130e2bd105dc8252ac/OMCompiler/SimulationRuntime/c/cmake_3.14.cmake#L12-L15

You have added both source files (normal and debug) but only one of them is intended to be compiled. Why is the debug version file there? How and when should it be compiled instead of the normal one.

My suggestion is to remove the debug file and move its additional features to the `newton_diagnostics.c` protected by some debug signifying macro define. The two files are very similar and there is no point in carrying duplicate code. You can use one of the existing debug flags (@AnHeuermann any debug flags/macros that would make sense for this) or you can just add a custom macro define just for this.

**Comment by AnHeuermann** [comment]
Theres already a debug flag added in [my initial commit](https://github.com/AnHeuermann/OpenModelica/commit/4e27b1105f667dbe75df6f58e177395fdb917990): `LOG_NLS_NEWTON_DIAG`.

That should be fine for debugging messages.

File: OMCompiler/SimulationRuntime/c/simulation/solver/newton_diagnostics.c [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+      id = 4;
+   else if (idx == 4)
+      id = 5;
+   else if (idx == 5)
+      id = 3;
+   return id;
+}*/
+// --------------------------------------------------------------------------------------------------------------------------------
+
+void MatMult( unsigned rA, unsigned cArB, unsigned cB, double A[rA][cArB], double B[cArB][cB], double C[rA][cB])
```
[the remark this answers is not in the document]
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/newton_diagnostics.c:134]: I guess this should be [DGEMM](http://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3_gaeda3cbd99c8fb834a60a6412878226e1.html)

**Review by AnHeuermann — CHANGES_REQUESTED** [review]

**Comment by AnHeuermann** [comment]
Hey @TeusvanderStelt. This PR seems to be stuck. Are you still working on this or should we close it for the time beeing?

**Comment by AnHeuermann** [comment]
Closing this for now. @TeusvanderStelt can open a new PR when he's ready.

**Pull request closed without merging**